[SYCL]Link Fallback Device Libraries On Demand

jinge90 opened the pull request:
Signed-off-by: gejin <ge.jin@intel.com>

jinge90 commented:
Hi, all. This patch implements the fallback device library(SPIRV files) on-demand loading.
The idea is:
1. After llvm-mc linking all kernel code into final device image, sycl-post-link tool will go through all modules to find all undefined "__devicelib_*" function calls.
2. For each undefined "__devicelib_*" functional call detected, we will check which fallback device library is required to support it.
3. After collecting all device libraries required, we will create an uint32_t "device library require mask" whose each bit corresponds to one fallback device library.
4. We store the "device library require mask" into "SYCL/devicelib req mask" property via sycl toolchain's "property" mechanism.
5. When sycl runtime loads device images, it will read the "device library require mask" and only load/link necessary fallback device libraries. If program doesn't use any device library functions, no fallback SPIRV file will be linked.
Hi, @vzakhari 
Could you help evaluate the potential risk in OMP's perspective, I expect no impact as sycl runtime is not used in OMP offloading.
Thank you very much.

asavonic commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -456,15 +537,23 @@ int main(int argc, char **argv) {
+          SpecConstantsPass::collectSpecConstantMetadata(
+              *MUptr.get(), ResultSpecIDMaps.back());
+      }
+      assert(DeviceLibReqMaskVec.size() == ResultSpecIDMaps.size());
```
Why this assertion is needed? It will crash if `(DoSpecConst && SetSpecConstAtRT) == false`, right?

jinge90 replied:
For each sub module, we will have one "require mask" stored in property file. And if (DoSpecConst && SetConstAtRt) == true, we will store one SpecIDMap for each sub module, so the number of "require mask" and "SpecIDMaps" should be the same.
If (DoSpecConst && SetSpecConstAtRT) == false, the assert will be skipped and we will not store SpecIDMaps in prop file for each submodule.

asavonic replied:
You're right, I got the indentation wrong.

vzakhari commented:
Hi @jinge90, you are right, there is no interaction with OpenMP currently.

Just curious, have you considered running dead code elimination in `sycl-post-link` so that it removes unused `devicelib` wrappers before propagating the property?

Can you please add header comments in `devicelib` files mentioning internal structures of `sycl-post-linlk` that need to be updated, if new `__devicelib_` APIs are added?  It is probably also worth updating `devicelib` documentation.

kbobrovs commented:
> After collecting all device libraries required, we will create an uint32_t "device library require mask" whose each bit corresponds to one fallback device library.

Why this is uint32_t bitmask rather than a list of  symbolic library ids like in ELF? If this is an optimization, I feel that it won't bring tangible speedup, but complicate code.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,15 +296,95 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
   return Res;
 }
 
+// Each fallback device library corresponds to one bit in "require mask" which
```
As I mentioned in the comments to the description, this additional mapping if library names to bits should better be avoided if there is no compelling performance benefit, as it obscures the logic.

kbobrovs replied:
I can see now that numeric identification of  libraries have existed before - the `DeviceLibExt` enum in program_manager.hpp. So I withdraw the above review comment - too much to rework.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,15 +296,95 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+  if (FuncName == "__devicelib_assert_fail") {
+    return DeviceLibAssert;
+  }
+  size_t Len =
```
suggest to factor out the repeated pattern into a function

jinge90 replied:
Hi, @kbobrovs 
The repeated patter has been removed since we don't use sorted array + binary_search any more.

kbobrovs commented on llvm/tools/sycl-post-link/DeviceLibFunctions.h:
```diff
@@ -0,0 +1,62 @@
+#ifndef __DEVICELIB_FUNCS_LIST__
+#define __DEVICELIB_FUNCS_LIST__
+#include <string>
+// All __devicelib_* functions must be sorted in alphabetic order as we
```
This looks like a source of obscure bugs. Suggest that this sorting requirement is dropped and the tool sorts automatically at startup. Or creates a map and uses map lookup instead of binary_search.

jinge90 replied:
Hi, @kbobrovs 
I have updated the patch to use map instead of sorted array.

andykaylor commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,15 +296,95 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+// 1. The function name has prefix "__devicelib_"
+// 2. The function has SPIR_FUNC calling convention
+// 3. The function is declaration which means it doesn't have function body
+static uint32_t getModuleReqMask(const std::unique_ptr<Module> &MPtr) {
```
Why are you using unique_ptr here and in getDeviceLibReqMasks?

jinge90 replied:
Hi, @andykaylor 
The main function in sycl-post-link.cpp has created a "std::vector<std::unique_ptr<Module>> ResultModules" and getDeviceLibReqMasks is invoked in main function, so I just use "ResultModules" as input parameters to avoid any conversion.
For getModuleReqMask, unique_ptr<Module> is not necessary, I have updated the patch to use "const Module &".

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,15 +296,95 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+// is an unsigned int32. getDeviceLibBit checks which fallback device library
+// is required for FuncName and returns the corresponding bit. The corresponding
+// mask for each fallback device library is:
+// fallback-cassert:      0x1
```
The library identification in program_manager.hpp and post link tool should match, so I suggest:
1) Add a cross-reference between sycl-post-link.cpp and program_manager.hpp and a comment that the library ids should match.
2) introduce `enum DeviceLibExt` exactly matching what's in program_manager.hpp here, so that grep over the sources gives all related occurrences.
3) Instead of individual definitions like `static constexpr uint32_t DeviceLib*` here use bit shift operations to translate e.g. `cl_intel_devicelib_math_fp64` to the corresponding bitmask's bit and back. Then we'll have to maintain single enum (in two places) rather than two inter-related.

andykaylor commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,15 +296,95 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+    std::map<StringRef, uint32_t> reqMaskEntry;
+    reqMaskEntry["devicelib_req_mask"] = ReqMaskVec[I];
+    PropSet.add(llvm::util::PropertySetRegistry::SYCL_DEVICELIB_REQ_MASK,
+                reqMaskEntry);
```
This always adds a map with a single entry?

jinge90 replied:
Yes, for each sub-module, we will add one mask.

kbobrovs commented on llvm/tools/sycl-post-link/DeviceLibFunctions.h:
```diff
@@ -0,0 +1,62 @@
+    "__devicelib_cpolar",   "__devicelib_cpow",     "__devicelib_cproj",
+    "__devicelib_creal",    "__devicelib_csin",     "__devicelib_csinh",
+    "__devicelib_csqrt",    "__devicelib_ctan",     "__devicelib_ctanh"};
+#endif
```
If devicelib_complex_fp64.spv changes the set of exported functions, which is quite likely, this place should be updated as well. So, please add a comment here and in the library sources. Otherwise people might have no idea why things are failing after they updated the .spv.

I'd suggest to also add a TODO to generate the list of functions automatically during the build, so that there is no need to update sycl-post-link every time device lib APIs change.

jinge90 replied:
All comments are added and since we have changed to use string map to store the devicelib info in sycl-post-link, the automation tool is not needed any more.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -456,15 +537,23 @@ int main(int argc, char **argv) {
     }
-    string_vector Files = saveSpecConstantIDMaps(ResultSpecIDMaps);
+    string_vector Files =
+        saveDeviceImageProperty(DeviceLibReqMaskVec, ResultSpecIDMaps);
```
I suggest to refactor property saving now, since with addition of the lib mask property (1) it is no longer scalable and the logic becomes too complex, IMO and (2) .prop file is now generated always.

e.g. 
```
  for (unsigned I = 0; I < ResultModules.size(); ++I) {
    std::string SCFile = makeResultFileName(".prop", I);
    llvm::util::PropertySetRegistry PropSet;
    std::error_code EC;
    raw_fd_ostream SCOut(SCFile, EC);

    if (DoSpecConst && SetSpecConstAtRT) {
      PropSet.add(llvm::util::PropertySetRegistry::SYCL_SPECIALIZATION_CONSTANTS,...);
      PropSet.write(SCOut);
    }
    if (needToAddLibMask) {
      PropSet.add(llvm::util::PropertySetRegistry::SYCL_LIB_MASK, ...);
      PropSet.write(SCOut);
    }
    if (needOtherFancyFeature) {
      PropSet.add(llvm::util::PropertySetRegistry::SYCL_OTHER_FANCY_FEATURE, ...);
      PropSet.write(SCOut);
    }
    Res.emplace_back(std::move(SCFile));
  }
```

jinge90 replied:
Hi, @kbobrovs 
I have refactored property saving in sycl-post-link, could you help review to see if the change aligns with your expectation?
Thank you very much.

jinge90 replied:
Hi, @kbobrovs 
Do you have any comments about the latest version of property saving function?
Thank you very much.

kbobrovs commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -659,15 +663,37 @@ RTDeviceBinaryImage &ProgramManager::getDeviceImage(OSModuleHandle M,
+  static constexpr uint32_t DeviceLibCmath = 0x2;
+  static constexpr uint32_t DeviceLibCmath64 = 0x4;
+  static constexpr uint32_t DeviceLibComplex = 0x8;
+  static constexpr uint32_t DeviceLibComplex64 = 0x10;
```
moreover, we don't really need them, and can use cl_intel_devicelib_assert,... + bit shifts to create masks

kbobrovs replied:
@andykaylor , is there a mechanism, BTW, to share definitions between the SYCL runtime and llvm tools? It then could be used for few other things as well.

andykaylor replied:
I don't think we've defined a formal mechanism. In general, the LLVM libraries are designed around a layering concept, which isn't documented but can be deduced from the cmake dependencies. When new classes or definitions are added, you need to find the right layer to put them in so that they can be shared among all the things that depend on them.

For a simple constant definition like this, it doesn't need to be complicated. Just add a header file at the lowest common level. Things like class definitions are more complicated because you have to make sure there are no circular dependencies or similar problems.

kbobrovs replied:
> Just add a header file at the lowest common level.

right, but what is this level between the llvm and runtime? Basically what we need here I guess is a mechanism to have a static library (or at least headers) which can be used both from llvm tools and SYCL runtime.

jinge90 replied:
Hi, @andykaylor  and @kbobrovs 
I think what you are suggesting is to put the the definition shared between llvm-tools and sycl runtime in a low level header and include it in both llvm tools and sycl runtime.
We may have different options:
1. created a new header in llvm/include/....
2. created a new header in sycl/include/CL/sycl...
Which one do you prefer or do you have other idea?
Thank you very much.

jinge90 replied:
Hi, @andykaylor  and @kbobrovs 
To share definition between llvm tools and sycl runtime, I added SYCLRTShared.h in llvm/Support and moved the definition of "enum DeviceLibExt" to this header file. This header file will be included in sycl-post-link.cpp and program_manager.hpp. Could you help me review it?
Thank you very much.

jinge90 commented:
> Hi @jinge90, you are right, there is no interaction with OpenMP currently.
> 
> Just curious, have you considered running dead code elimination in `sycl-post-link` so that it removes unused `devicelib` wrappers before propagating the property?
> 
> Can you please add header comments in `devicelib` files mentioning internal structures of `sycl-post-linlk` that need to be updated, if new `__devicelib_` APIs are added? It is probably also worth updating `devicelib` documentation.

Hi, @vzakhari , I have considered dead code elimination and drafted a simple patch in sycl-post-link. I will send the patch to you later.

jinge90 commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,15 +296,95 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+  static constexpr uint32_t DeviceLibCmath64 = 0x4;
+  static constexpr uint32_t DeviceLibComplex = 0x8;
+  static constexpr uint32_t DeviceLibComplex64 = 0x10;
+  if (FuncName == "__devicelib_assert_fail") {
```
Yes, a string map can simplify the implementation, already updated the patch.

jinge90 commented on llvm/tools/sycl-post-link/DeviceLibFunctions.h:
```diff
@@ -0,0 +1,62 @@
+#include <string>
+// All __devicelib_* functions must be sorted in alphabetic order as we
+// will use binary search to find some entry in them.
+static std::string CmathDeviceLibFunctions[] = {
```
Already updated the patch to move all array definition to .cpp file.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,20 +438,73 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+  // perf regression is observed if we don't link any device library. The perf
+  // regression is caused by a clang issue.
+  uint32_t ReqMask = 0x1;
+  uint32_t DeviceLibBits = 0;
```
please move this declaration inside the scope where it is used

jinge90 replied:
Done.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,20 +438,73 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+// 2. The function has SPIR_FUNC calling convention
+// 3. The function is declaration which means it doesn't have function body
+static uint32_t getModuleReqMask(const Module &M) {
+  // 0x1 means sycl runtime will link and load libsycl-fallback-assert.spv as
```
Suggest to check Module target triple being spirv instead of filtering by calling convention below.

jinge90 replied:
Hi, @kbobrovs 
I added check for target triple in this function, the logic is we will return "0"(means no devicelib is needed) if target triple is NOT spirv. But I am not sure whether I used the correct way to check the target triple string, I assumed that for spirv target, the target triple string should have the prefix "spir64-", is it a valid assumption?
Thank you very much.

kbobrovs replied:
I suggest `llvm::Triple(M.getTargetTriple()).isSPIR()`

jinge90 replied:
Done.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,20 +438,73 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+    const ImagePropSaveInfo &ImgPSInfo) {
+  string_vector Res;
+  for (size_t I = 0; I < ResultModules.size(); ++I) {
     std::string SCFile = makeResultFileName(".prop", I);
```
can be moved below to the place of actual use

jinge90 replied:
Hi, @kbobrovs 
What did you suggest to move below?

kbobrovs replied:
added suggested change below

jinge90 replied:
Done.

kbobrovs commented on sycl/include/CL/sycl/detail/pi.h:
```diff
@@ -649,6 +649,7 @@ static const uint8_t PI_DEVICE_BINARY_OFFLOAD_KIND_SYCL = 4;
 /// PropertySetRegistry::SYCL_SPECIALIZATION_CONSTANTS defined in
 /// PropertySetIO.h
 #define PI_PROPERTY_SET_SPEC_CONST_MAP "SYCL/specialization constants"
+#define PI_PROPERTY_SET_DEVICELIB_REQ_MASK "SYCL/devicelib req mask"
```
since you introduced LLVM/RT shared header, these two declarations can be moved there

jinge90 replied:
Done.

kbobrovs commented on sycl/test/program_manager/env_vars.cpp:
```diff
@@ -5,15 +5,12 @@
 // RUN: %CPU_RUN_PLACEHOLDER SYCL_PROGRAM_COMPILE_OPTIONS="-g" %t.out
 // RUN: %GPU_RUN_PLACEHOLDER SYCL_PROGRAM_COMPILE_OPTIONS="-g" %t.out
-// RUN: %CPU_RUN_PLACEHOLDER SYCL_PROGRAM_LINK_OPTIONS="-enable-link-options -cl-denorms-are-zero" %t.out
-// RUN: %GPU_RUN_PLACEHOLDER SYCL_PROGRAM_LINK_OPTIONS="-enable-link-options -cl-denorms-are-zero" %t.out
```
why is this removed?

jinge90 replied:
Hi, @kbobrovs 
Currently, all program's kernel code will be linked with our devicelib as default, so sycl runtime will use piProgramLink to link all spv images. However, after we enabled devicelib "on-demand" loading,  sycl runtime will use piProgramBuild to build users' device image. For OCL backend, piProgramBuild will be mapped to OCL's clBuildProgram API. According to OCL Spec, we can only pass "Compiler options" to this API and "Link Options" is not used.
To this test, the source code doesn't depend on devicelib, so I remove all test commands using "SYCL_PROGRAM_LINK_OPTIONS".
However, current status is a little tricky, we have  to keep loading "assert" fallback spv as default to workaround another separate issue, so sycl runtime will still use piProgramCompile + piProgramLink for all program's device code. Do you think we need to keep those test commands for now?
Thank you very much.

kbobrovs replied:
I think the problems you describe are implementation detail and should be fixed. User setting `SYCL_PROGRAM_LINK_OPTIONS=` should not lead to crashes if link options are good. If that's the case, the implementation should be fixed, I think.

jinge90 replied:
Hi, @kbobrovs 
For this case, those link options will lead to crash in OCL CPU runtime, the problem is in "-enable-link-options". If no device libraries will be loaded and linked with user's kernel code, sycl runtime will use piProgramBuild for jit compilation which will be mapped to clBuildProgram. According to OCL Spec, "-enable-link-options" must be specified with "--create-library" option.  I think we have 2 ways to deal with the case:
1. If sycl runtime uses piProgramBuild, the SYCL_PROGRAM_LINK_OPTIONS env variable value will be ignored, we only pass SYCL_PROGRAM_COMPILER_OPTIONS value to underlying OCL runtime.
2. We keep passing "SYCL_PROGRAM_LINK_OPTIONS" to underlying OCL runtime even if piProgramBuild is used, it's user's duty to set "good" SYCL_PROGRAM_LINK_OPTIONS. We need to fix this case to remove "-enable-link-options".
Which one do you prefer or do you have any other idea?
Thank you very much.

kbobrovs replied:
> According to OCL Spec, "-enable-link-options" must be specified with "--create-library" option.

I see, thanks. This makes the test invalid, so removal is OK then.
2) seems better solution. But for this PR just removal of the two invalid test cases (as you did before) should be enough

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,20 +438,73 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+  uint32_t DeviceLibBits = 0;
+  for (const Function &SF : M) {
+    if (SF.getName().startswith(DEVICELIB_FUNC_PREFIX) &&
+        (SF.getCallingConv() == CallingConv::SPIR_FUNC) && SF.isDeclaration()) {
```
`SF.getCallingConv() == CallingConv::SPIR_FUNC`
this should really be an assert as we don't really expect non-spirv functions with matching prefix

jinge90 replied:
Hi, @kbobrovs 
I agree that we don't expect any non-spirv functions with matching prefix, need we add a note in document telling users that __devicelib_* names are used by DPC++ compiler and they can't use those names in their code?
Thank you very much.

kbobrovs replied:
Users can't define names starting with "__" anyway - those are reserved for system/compiler, so I don't see a problem. Adding to the doc won't hurt, though :)

jinge90 replied:
OK, got it.
Already updated the patch to use assert for (SF.getCallingConv() == CallingConv::SPIR_FUNC).

andykaylor commented on llvm/include/llvm/Support/SYCLRTShared.h:
```diff
@@ -0,0 +1,23 @@
+//=- SYCLRTShared.h - Shared definition between llvm tools and SYCL runtime -=//
```
This seems OK to me.

jinge90 replied:
Hi, @kbobrovs 
I have added the comments in SYCLRTShared.h.

bader replied:
@kbobrovs, thanks for pinging me and sorry for late feedback.

I don't fully understand the design of this "linking on demand" solution (it would be great to extend https://github.com/intel/llvm/blob/sycl/sycl/doc/CompilerAndRuntimeDesign.md document), but this concerns me.
It seems like we bake into LLVM compiler framework the information about Intel specific OpenCL extensions. @andykaylor, do you think this design can be accepted by the LLVM community? I think community prefers scaleable solutions (e.g. framework for linking arbitrary libraries), whereas current implementation is able to address very specific SYCL use case.

It would be great to get a community feedback on this approach through RFC.

kbobrovs replied:
> It seems like we bake into LLVM compiler framework the information about Intel specific OpenCL extensions.
@bader, under  "Intel specific OpenCL extensions" do you mean SYCL as a whole or just the "linking on demand part"?
If the former (SYCL), then in my view we already have baked a lot of SYCL into LLVM :) (for good reason) - take clang FE, for example, offload tools - bundler, wrapper, SYCL specific passes in LLVM itself. Maybe the location "llvm/Support/SYCLRTShared.h" is not ideal and there could be new folder with "SYCL" in it to underline SYCL specifics?

If the latter - then SYCLRTShared.h is not specific to this extension, it serves for sharing some type and string literal definitions between LLVM tools and SYCL runtime.

> It would be great to get a community feedback on this approach through RFC.
Do you think this should be done before this PR can be merged or can be done in parallel?

bader replied:
> > It seems like we bake into LLVM compiler framework the information about Intel specific OpenCL extensions.
> 
> @bader, under "Intel specific OpenCL extensions" do you mean SYCL as a whole or just the "linking on demand part"?

cl_intel_devicelib_assert, cl_intel_devicelib_math, cl_intel_devicelib_math_fp64, cl_intel_devicelib_complex, cl_intel_devicelib_complex_fp64 - are Intel specific OpenCL extensions.

> If the former (SYCL), then in my view we already have baked a lot of SYCL into LLVM :) (for good reason) - take clang FE, for example, offload tools - bundler, wrapper, SYCL specific passes in LLVM itself. Maybe the location "llvm/Support/SYCLRTShared.h" is not ideal and there could be new folder with "SYCL" in it to underline SYCL specifics?
> 
> If the latter - then SYCLRTShared.h is not specific to this extension, it serves for sharing some type and string literal definitions between LLVM tools and SYCL runtime.
> 

Can we apply the same approach as regular C++ compiler does? I.e. link with the whole C++ standard library (using LLVM linker) + provide an compiler option to disable [SYCL extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst) and skip linking with this libraries? This should much easier to implement and maintain as it should not require extending LLVM functionality.

> > It would be great to get a community feedback on this approach through RFC.
> 
> Do you think this should be done before this PR can be merged or can be done in parallel?

Ideally the review should happen before the merge, but if there are any time constraints we can merge and re-do later if needed.

jinge90 replied:
Hi, @kbobrovs @andykaylor @bader 
This patch is used to fix some perf regression for cases which has "small" kernel execution time and we are required to fix it ASAP recently, could we merge it and continue to work on the RFC to collect community's feedback?
And I think it is valuable to add a header to share information between sycl runtime and llvm-tools. Currently, the header is located in llvm/Support/SYCLRTShared.h and following things are included:
1. Property constant string such as "SYCL/specialization constants".
2. DeviceLibExt enum definition such as "cl_intel_devicelib_***"
We need a careful analysis and discussion on following points:
1.  The header's filename and location?
2.  The namespace used in this header?
3.  Which  definitions or constants can be placed in this file?
4. Can intel-specific things be placed in this header?
Thank you very much.

bader replied:
I'm not convinced that this is the right approach to "fix performance regression". Moving some runtime overhead to compile time (e.g. https://github.com/intel/llvm/pull/1398) seems to be a better option.
What is the root cause of the regression? Please, add a problem statement to the PR description.

jinge90 commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,20 +438,73 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+    return 0;
+  else
+    return 0x1 << (DeviceLibFuncMap[FuncName] - cl_intel_devicelib_assert);
+}
```
Done.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,20 +438,74 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+// 3. The function is declaration which means it doesn't have function body
+static uint32_t getModuleReqMask(const Module &M) {
+  // Device libraries will be enabled only for spir-v module.
+  if (M.getTargetTriple().substr(0, 7) != "spir64-")
```
this can crash if triple is too short. .startswith would work better, but see below.

jinge90 replied:
Done.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,20 +438,74 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+    const ImagePropSaveInfo &ImgPSInfo) {
+  string_vector Res;
+  for (size_t I = 0; I < ResultModules.size(); ++I) {
     std::string SCFile = makeResultFileName(".prop", I);
```
```suggestion
```

jinge90 replied:
Done.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,20 +438,74 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+          TmpSpecIDMap);
+    }
     std::error_code EC;
     raw_fd_ostream SCOut(SCFile, EC);
```
```suggestion
    std::string SCFile = makeResultFileName(".prop", I);
    raw_fd_ostream SCOut(SCFile, EC);
```

jinge90 replied:
Done.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -295,20 +438,74 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
     PropSet.write(SCOut);
     Res.emplace_back(std::move(SCFile));
   }
+
```
```suggestion
```

jinge90 replied:
Done.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -456,18 +652,15 @@ int main(int argc, char **argv) {
     Error Err = Table.addColumn(COL_PROPS, Files);
     CHECK_AND_EXIT(Err);
   }
+
```
```suggestion
```

jinge90 replied:
Done

kbobrovs commented on llvm/include/llvm/Support/SYCLRTShared.h:
```diff
@@ -0,0 +1,29 @@
+/// Name must be consistent with
+/// PropertySetRegistry::SYCL_SPECIALIZATION_CONSTANTS defined in
+/// PropertySetIO.h
+#define PI_PROPERTY_SET_SPEC_CONST_MAP "SYCL/specialization constants"
+#define PI_PROPERTY_SET_DEVICELIB_REQ_MASK "SYCL/devicelib req mask"
```
```suggestion
#define SYCL_PROPERTY_SET_SPEC_CONST_MAP "SYCL/specialization constants"
#define SYCL_PROPERTY_SET_DEVICELIB_REQ_MASK "SYCL/devicelib req mask"
```

kbobrovs replied:
Thanks for moving it here. Now, it should be removed from PropertySetIO.[cpp|h] too, and all sources (post-link tool and SYCL runtime) which use that should include this header.

jinge90 replied:
Done.

kbobrovs commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -123,6 +123,7 @@ class PropertySetRegistry {
   // Specific property category names used by tools.
   static constexpr char SYCL_SPECIALIZATION_CONSTANTS[] =
       "SYCL/specialization constants";
+  static constexpr char SYCL_DEVICELIB_REQ_MASK[] = "SYCL/devicelib req mask";
```
```suggestion
```

kbobrovs replied:
this should be moved to your new RT-LLVM shared header

jinge90 replied:
Done.

kbobrovs commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -36,6 +37,11 @@
 #include <memory>
 
 using namespace llvm;
+using llvm::util::sycl::cl_intel_devicelib_assert;
+using llvm::util::sycl::cl_intel_devicelib_complex;
+using llvm::util::sycl::cl_intel_devicelib_complex_fp64;
+using llvm::util::sycl::cl_intel_devicelib_math;
+using llvm::util::sycl::cl_intel_devicelib_math_fp64;
```
somehow my previous comment on this disappeared. Please don't `using` all the individual enum elements.

kbobrovs replied:
BTW, I see that you add merge commits when you implement review comments:
> Merge remote-tracking branch 'upstream/sycl' into link_fallback_devic… …

This does not seem right to me. Maybe @bader can comment more. I think this can lead to missing review comments. You should only use git push/git pull when working with the review, so that `jinge90:link_fallback_devicelib_on_demand` does not have merge commits. Only if you have conflicts, you should rebase your feature branch. I do it like this:

`git rebase <hash> --onto intel_llvm/sycl`

where <hash> is the commit right below the first feature branch commit in `git log` output.

jinge90 replied:
Hi, @kbobrovs 
Thank you for the remind.  I updated patch to use c++11 scoped enum for DeviceLibExt, the definition looks like:
enum class DeviceLibExt : std::uint32_t {cl_intel_devicelib_assert,.....}
By doing this, we don't need to "using" all individual enum elements, we only need to "using llvm::util::sycl::DeviceLibExt;" in source code and use "DeviceLibExt::cl_intel_devicelib_*" to refer to all DeviceLibExt enum elements. Is this OK to you?

kbobrovs replied:
Yes, thank you.

kbobrovs commented on sycl/source/detail/program_manager/program_manager.hpp:
```diff
@@ -44,13 +44,12 @@ class context_impl;
-  cl_intel_devicelib_complex,
-  cl_intel_devicelib_complex_fp64
-};
+using llvm::util::sycl::cl_intel_devicelib_assert;
+using llvm::util::sycl::cl_intel_devicelib_complex;
+using llvm::util::sycl::cl_intel_devicelib_complex_fp64;
+using llvm::util::sycl::cl_intel_devicelib_math;
+using llvm::util::sycl::cl_intel_devicelib_math_fp64;
+using llvm::util::sycl::DeviceLibExt;
```
please don't `using` individual enum elements

bader commented:
@smaslov-intel, @intel/llvm-reviewers-runtime, ping.

jinge90 commented:
> I'm okay to merge if someone is waiting for this fix. I'd like to have separate discussion to better understand the problem this PR is supposed to solve, because I have serious concerns that we won't be able to upstream this solution.

Hi, @bader @kbobrovs @andykaylor 
"devicelib on-demand loading" is an optimization for current devicelib infra. And we can implement it without introducing header file to llvm/Support/...(We didn't touch any thing in llvm/ in the original patch)  In fact, introducing some mechanism or header files to LLVM framework to share information between SYCL runtime and llvm tools  is another separate problem and this problem is much more general than this patch. I agree with Alexey that we need a separate discussion on this problem.
Thank you very much,

kbobrovs commented:
@jinge90, I'm ok to revert the shared header and return to code duplication if there is no consensus on the solution. If you do this, please restore cross-reference in the comments between the duplicated parts and match naming.

jinge90 commented:
> @jinge90, I'm ok to revert the shared header and return to code duplication if there is no consensus on the solution. If you do this, please restore cross-reference in the comments between the duplicated parts and match naming.

Hi, @kbobrovs 
I have reverted the shared header and added the comments to remind sync between sycl runtime and sycl-post-link tool.
Thank you very much.

jinge90 commented:
Hi, @kbobrovs @bader @asavonic 
I have removed the SYCLRTShared.h to avoid touching LLVM framework, could you help approve again.
Thank you very much.
